integrationtest/backend/lucene/src/test/java/org/hibernate/search/integrationtest/backend/lucene/mapping/LuceneVectorFieldIT.java

```diff
@@ -54,4 +68,166 @@ void test(int dimension, int efConstruction, int m, String property, int value,
+
+		setupHelper.start().withIndex( index ).setup();
+
+		index.bulkIndexer()
+				.add( "ID:1", document -> {
+					document.addValue( "vector", vector );
+				} )
+				.join();
+
+		assertThat(
+				index.createScope().query().select( f -> f.field( "vector", float[].class ) )
+						.where( f -> f.knn( 1 ).field( "vector" ).matching( vector ) )
+						.fetchAllHits()
+		).hasSize( 1 )
+				.contains( foundVector );
+
+		float[] badVector = Arrays.copyOf( vector, vector.length );
+		badVector[0] += 50;
+
+		assertThat(
+				index.createScope().query().select( f -> f.field( "vector", float[].class ) )
+						.where( f -> f.knn( 1 ).field( "vector" ).matching( badVector ) )
+						.fetchAllHits()
+		).hasSize( 1 )
+				.contains( foundVector );
+	}
+
+	private static List<? extends Arguments> floatVectors() {
+		return List.of(
+				Arguments.of( VectorSimilarity.L2, 5.0f, new float[] { 5.0f, 0, 0, 0, 0 }, new float[] { 5.0f, 0, 0, 0, 0 } ),
+				Arguments.of( VectorSimilarity.L2, 5.0f, new float[] { 0, 5.0f, 0, 0, 0 }, new float[] { 0, 5.0f, 0, 0, 0 } ),
+				Arguments.of( VectorSimilarity.L2, 4.0f, new float[] { 2.0f, 2.0f, 2.0f, 2.0f, 0 },
+						new float[] { 2.0f, 2.0f, 2.0f, 2.0f, 0 } ),
+				Arguments.of( VectorSimilarity.L2, 5.0f, new float[] { 25.0f, 0, 0, 0, 0 }, new float[] { 5.0f, 0, 0, 0, 0 } ),
+				Arguments.of( VectorSimilarity.DOT_PRODUCT, 16.0f, new float[] { 2.0f, 2.0f, 2.0f, 2.0f, 0 },
+						new float[] { 2.0f, 2.0f, 2.0f, 2.0f, 0 } ),
+				Arguments.of( VectorSimilarity.DOT_PRODUCT, 25.0f, new float[] { 5.0f, 0, 0, 0, 0 },
+						new float[] { 5.0f, 0, 0, 0, 0 } ),
+				Arguments.of( VectorSimilarity.DOT_PRODUCT, 155.0f, new float[] { 5.0f, 0, 0, 0, 0 },
+						new float[] { 31.0f, 0, 0, 0, 0 } )
+		);
+	}
+
+	@Test
+	void notSureThatItWouldBeOkToNormalizeByteVectorsSeeThisExampleWhy() {
+		byte[] vector = { 5, 1, 1, 1, 1 };
+		int dotProduct = VectorUtil.dotProduct( vector, vector );
+		double l2norm = Math.sqrt( dotProduct );
+
+		float[] vector2 = new float[vector.length];
+
+		float magnitude = 120.0f;
+
+		for ( int i = 0; i < vector2.length; i++ ) {
+			vector2[i] = (float) ( vector[i] * magnitude / l2norm );
+		}
+
+		double dotProduct2 = VectorUtil.dotProduct( vector2, vector2 );
+		double l2norm2 = Math.sqrt( dotProduct2 );
+		assertThat( l2norm2 ).isCloseTo( magnitude, Percentage.withPercentage( 0.0001 ) );
+		assertThat( vector2 ).contains(
+				111.417206f,
+				22.28344f,
+				22.28344f,
+				22.28344f,
+				22.28344f
+		);
+
+		byte[] vector3 = new byte[vector.length];
+
+		for ( int i = 0; i < vector2.length; i++ ) {
+			vector3[i] = (byte) ( vector[i] * magnitude / l2norm );
+		}
+
+		double dotProduct3 = VectorUtil.dotProduct( vector3, vector3 );
+		double l2norm3 = Math.sqrt( dotProduct3 );
+		// l2norm3 == 119.4026800369238
+		assertThat( l2norm3 ).isCloseTo( magnitude, Percentage.withPercentage( 1 ) );
+		assertThat( vector3 ).contains(
+				111,
+				22,
+				22,
+				22,
+				22
+		);
+	}
```

Comment:
This is why I'm not "happy" to normalize byte[] vectors ...  (that is if my understanding of it is correct)
because of the precision/data loss, we won't get the norm value we expect ...

Reply:
I understand, but I think we need to answer this before we can take a decision:

1. Is there anything about normalizing byte vectors in OpenSearch?
2. For similarities that require normalized vectors, what magnitude do they require for byte vectors?
3. Does requring a magnitude different from whatever they require make sense in practice, and if not do we have the same problem for that magnitude? E.g. we can certainly ensure that all byte vectors have a magnitude of 255, since a `byte` magnitude cannot be more than 255, so rounding all coordinates **up** (instead of down like in this example) would do the trick.

Reply:
> dot_product
Computes the dot product of two unit vectors. This option provides an optimized way to perform cosine similarity. 
>
> When element_type is float, all vectors must be unit length.
>
> When element_type is byte, all vectors must have the same length.

^ from what I see only `dot_product` requires vectors to be normalized. (and cosine cannot accept a vector ith zero magnitude)

> Is there anything about normalizing byte vectors in OpenSearch?

I haven't seen anything, but since OpenSearch only accepts `L2` and `Consine` (https://github.com/opensearch-project/k-NN/blob/e1a7cb7aa1a523cf75beba3446985a87682f4d2c/src/main/java/org/opensearch/knn/index/util/Lucene.java#L45) with a Lucene engine, it doesn't really need it ... 

As for Elasticsearch: it has a check for floats:

```
The [dot_product] similarity can only be used with unit-length vectors. Preview of invalid vector: [1.0, 2.0, 3.0]
```
but I can put any byte vectors in with no exceptions...

which matches the checks here : 

https://github.com/elastic/elasticsearch/blob/eeecdbf87b08ec03682dabe7def7272055119252/server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java#L322-L338 

> For similarities that require normalized vectors, what magnitude do they require for byte vectors?

so from that `dot_product` description `all vectors must have the same length`... my understanding is that it can be any length e.g. 23 😃 , but just all of them should be of the same length...


When I'm thinking about it, I'd expect that the model used to produce the vector should provide some guidance here, i.e. does it produce normalized vectors or not and based on that user decides what to do.. IDK if it makes sense to transform the vectors produced by the model...

Reply:
> When I'm thinking about it, I'd expect that the model used to produce the vector should provide some guidance here, i.e. does it produce normalized vectors or not and based on that user decides what to do.. IDK if it makes sense to transform the vectors produced by the model...

Makes sense... in that case checks would be enough (i.e. HSEARCH-5048 but not HSEARCH-5019)... but! We need an expected magnitude, or we cannot run checks.

From what you say here, I understand that with some similarities at least, we can infer the expected magnitude:

> As for Elasticsearch: it has a check for floats:
> 
> ```
> The [dot_product] similarity can only be used with unit-length vectors. Preview of invalid vector: [1.0, 2.0, 3.0]
> ```

So to sum up:

* Normalization probably doesn't really make sense, so we'd skip that (close HSEARCH-5019)
* Checks may make sense so we'd keep them (HSEARCH-5048)
* We'd enable checks by default depending on the configured similarity -- I think only for `dot_product` with `float` vectors?
* We may want to provide a way to disable checks (see other conversation about an enum)
* We may want to keep the custom magnitude and a way to enable checks explicitly, but we're not sure that really makes sense, so maybe we shouldn't.

Reply:
Thanks! I've started to look into the checking part .... 

> We may want to provide a way to disable checks (see other conversation about an enum)

I guess we won't be able to disable it for elasticsearch backends ... as it will fail on the server side, so maybe start with just adding a check to the Lucene impl and see if users request a change, oooor add a Lucene-specific backend config property to disable the checks?

Reply:
I was more thinking of people who have normalized vectors anyway, so don't want to waste time on checks that are not useful.

But ok, +1 to not allow disabling it for a start.

Implementing it for the Elasticsearch backend would still have value, since the error message will probably be clearer, but if it's harder (e.g. you need to rely on Lucene classes), ok, let's not bother.

Reply:
no no, it should be fine to implement for Elastcisearch too, I was just thinking that for Elasticsearch we'd be doing the check twice (the second one happening on the server side), and if a user disables it on our side, Elasticsearch will still fail ...
maybe we just need to come up with a name that tells the user that it happens before we send the data to backend? but I don't have a good example of such name 🙈

Reply:
> ok, let's not bother.

:)

Reply:
https://github.com/hibernate/hibernate-search/pull/4022